core/src/main/java/org/apache/iceberg/DefaultModifyFiles.java

```diff
@@ -0,0 +1,114 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.expressions.Evaluator;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.Projections;
+
+public class DefaultModifyFiles extends MergingSnapshotProducer<ModifyFiles> implements ModifyFiles {
+
+  private Long baseSnapshotId;
+  private Expression rowFilter;
+
+  DefaultModifyFiles(TableOperations ops, Long baseSnapshotId) {
+    super(ops);
+    this.baseSnapshotId = baseSnapshotId;
+    // modify files must fail if any of the deleted paths is missing and cannot be deleted
+    failMissingDeletePaths();
+  }
+
+  @Override
+  protected ModifyFiles self() {
+    return this;
+  }
+
+  @Override
+  protected String operation() {
+    return DataOperations.MODIFY;
+  }
+
+  @Override
+  public ModifyFiles modifyFiles(Set<DataFile> filesToDelete, Set<DataFile> filesToAdd) {
+    Preconditions.checkArgument(filesToDelete != null, "Files to delete cannot be null");
+    Preconditions.checkArgument(filesToAdd != null, "Files to add cannot be null");
+
+    for (DataFile toDelete : filesToDelete) {
+      delete(toDelete.path());
+    }
+
+    for (DataFile toAdd : filesToAdd) {
+      add(toAdd);
+    }
+
+    return this;
+  }
+
+  @Override
+  public ModifyFiles failOnNewFiles(Expression newRowFilter) {
+    Preconditions.checkArgument(newRowFilter != null, "Row filter cannot be null");
+    this.rowFilter = newRowFilter;
+    return this;
+  }
+
+  @Override
+  public List<ManifestFile> apply(TableMetadata base) {
+    if (rowFilter != null) {
+      PartitionSpec spec = writeSpec();
+      Expression inclusiveExpr = Projections.inclusive(spec).project(rowFilter);
+      Evaluator inclusive = new Evaluator(spec.partitionType(), inclusiveExpr);
+
+      List<DataFile> newFiles = collectNewFiles(base);
+      for (DataFile newFile : newFiles) {
+        // we do partition-level conflict resolution right now
+        // we can enhance it by leveraging column stats and MetricsEvaluator
+        ValidationException.check(!inclusive.eval(newFile.partition()),
+            "Modify operation detected a new file %s that might match %s", newFile.path(), rowFilter);
+      }
+    }
+
+    return super.apply(base);
+  }
+
+  private List<DataFile> collectNewFiles(TableMetadata meta) {
+    List<DataFile> newFiles = new ArrayList<>();
+
+    Long currentSnapshotId = meta.currentSnapshot() == null ? null : meta.currentSnapshot().snapshotId();
+    while (currentSnapshotId != null && !currentSnapshotId.equals(baseSnapshotId)) {
+      Snapshot currentSnapshot = meta.snapshot(currentSnapshotId);
+
+      if (currentSnapshot == null) {
+        throw new ValidationException(
+            "Modify operation cannot find snapshot %d. Was it expired?", currentSnapshotId);
+      }
+
+      Iterables.addAll(newFiles, currentSnapshot.addedFiles());
```

Comment:
`currentSnapshot.addedFiles()` doesn't return stats, so it will complicate stats-based conflict resolution.

Reply:
We can go back to returning stats for this method.

Reply:
I'll do that in a follow-up PR.